nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,232 @@
+      Sep 2014, Boston, United States. pp.8
+    """
+    def __init__(self, n_regions=None, min_size=20,
+                 threshold='auto', percentile=80,
+                 extractor='autoextractor', seed_ratio='auto',
+                 smooth_fwhm=6., verbose=0):
+        self.n_regions = n_regions
+        self.min_size = min_size
+        self.threshold = threshold
+        self.percentile = percentile
+        self.extractor = extractor
+        self.seed_ratio = seed_ratio
+        self.smooth_fwhm = smooth_fwhm
+        self.verbose = verbose
+
+    def transform(self, maps_img):
+        """ Extract the regions from the maps
+
+        Parameters
+        ----------
+        maps_img: a Nifti like image/object or the filename
+            the image consists of atlas maps or statistical maps
+
+        Returns
+        -------
+        regions_: 4D Nifti like image/object
+            a list of seperate regions concatenated in a 4D Nifti
+            like image.
+
+        index_: a numpy array
+            a list of indices in an array to get the identity of each
+            region to a family of brain atlas maps.
+        """
+        min_size = self.min_size
+        maps_img = check_niimg(maps_img)
+
+        extractor_methods = ['autoextractor', 'randomwalker']
+        if (not isinstance(self.extractor, _basestring) or
+                self.extractor not in extractor_methods):
+            message = ('"extractor" should be given '
+                       'either of these {0}').format(extractor_methods)
+            raise ValueError(message)
+
+        if isinstance(self.threshold, float):
+            threshold = self.threshold
+        elif self.threshold == 'auto':
+            pass
+        elif self.threshold is not None:
+            raise ValueError("Threshold must be done, "
+                             "'auto' or float. You provided %s." %
+                             str(self.threshold))
+
+        regions_imgs = []
+        index = []
+        percentile = self.percentile
+
+        if self.extractor == 'randomwalker':
+            randomwalker = True
+            if isinstance(self.seed_ratio, float):
+                seed_ratio = self.seed_ratio
+            elif self.seed_ratio == 'auto':
+                pass
+            elif self.seed_ratio is not None:
+                raise ValueError("seeds_ratio must be, "
+                                 "'auto' or float. You have given %s." %
+                                 str(self.seed_ratio))
+        else:
+            randomwalker = False
+
+        for i, img in enumerate(iter_img(maps_img)):
+            data = img.get_data()
+            affine = img.get_affine()
+            each_map = data.copy()
+
+            if self.threshold == 'auto':
+                threshold = fast_abs_percentile(each_map[each_map != 0].ravel(),
+                                                percentile)
```

Comment:
Well, problems are meant to be solved ;). I gave you several options to solve them.
Your solution here is a workaround to aboid the fact that you need the brain mask: you suppose that nonzero values are the only one not masked, which is wrong.